go 1.12 - wasm

I just updated to 1.12 and when compiling my project with tinygo, I get the following error:

```
/usr/local/go/src/internal/fmtsort/sort.go:58:19: invalid operation: mapValue (variable of type reflect.Value) has no field or method MapRange
```

It compiled just fine with 1.11.5 earlier

--- a/compiler/compiler.go
+++ b/compiler/compiler.go
@@ -500,6 +500,16 @@ func (c *Compiler) getLLVMType(goType types.Type) (llvm.Type, error) {
 			members[i] = member
 		}
 		return c.ctx.StructType(members, false), nil
+	case *types.Tuple:
+		members := make([]llvm.Type, typ.Len())
+		for i := 0; i < typ.Len(); i++ {
+			member, err := c.getLLVMType(typ.At(i).Type())
+			if err != nil {
+				return llvm.Type{}, err
+			}
+			members[i] = member
+		}
+		return c.ctx.StructType(members, false), nil
 	default:
 		return llvm.Type{}, errors.New("todo: unknown type: " + goType.String())
 	}
@@ -1821,6 +1831,29 @@ func (c *Compiler) parseExpr(frame *Frame, expr ssa.Value) (llvm.Value, error) {
 		}
 		c.builder.CreateStore(zero, it)
 		return it, nil
+	case *ssa.Select:
+		if len(expr.States) == 0 {
+			// Shortcuts for some simple selects.
+			llvmType, err := c.getLLVMType(expr.Type())
+			if err != nil {
+				return llvm.Value{}, err
+			}
+			if expr.Blocking {
+				// Blocks forever:
+				//     select {}
+				c.createRuntimeCall("deadlockStub", nil, "")
+				return llvm.Undef(llvmType), nil
+			} else {
+				// No-op:
+				//     select {
+				//     default:
+				//     }
+				retval := llvm.Undef(llvmType)
+				retval = c.builder.CreateInsertValue(retval, llvm.ConstInt(c.intType, 0xffffffffffffffff, true), 0, "")
+				return retval, nil // {-1, false}
+			}
+		}
+		return llvm.Value{}, c.makeError(expr.Pos(), "unimplemented: "+expr.String())
 	case *ssa.Slice:
 		if expr.Max != nil {
 			return llvm.Value{}, c.makeError(expr.Pos(), "todo: full slice expressions (with max): "+expr.Type().String())

--- a/compiler/goroutine-lowering.go
+++ b/compiler/goroutine-lowering.go
@@ -170,6 +170,10 @@ func (c *Compiler) markAsyncFunctions() (needsScheduler bool, err error) {
 	if !sleep.IsNil() {
 		worklist = append(worklist, sleep)
 	}
+	deadlockStub := c.mod.NamedFunction("runtime.deadlockStub")
+	if !deadlockStub.IsNil() {
+		worklist = append(worklist, deadlockStub)
+	}
 	chanSendStub := c.mod.NamedFunction("runtime.chanSendStub")
 	if !chanSendStub.IsNil() {
 		worklist = append(worklist, chanSendStub)
@@ -288,7 +292,7 @@ func (c *Compiler) markAsyncFunctions() (needsScheduler bool, err error) {
 
 	// Transform all async functions into coroutines.
 	for _, f := range asyncList {
-		if f == sleep || f == chanSendStub || f == chanRecvStub {
+		if f == sleep || f == deadlockStub || f == chanSendStub || f == chanRecvStub {
 			continue
 		}
 
@@ -305,7 +309,7 @@ func (c *Compiler) markAsyncFunctions() (needsScheduler bool, err error) {
 			for inst := bb.FirstInstruction(); !inst.IsNil(); inst = llvm.NextInstruction(inst) {
 				if !inst.IsACallInst().IsNil() {
 					callee := inst.CalledValue()
-					if _, ok := asyncFuncs[callee]; !ok || callee == sleep || callee == chanSendStub || callee == chanRecvStub {
+					if _, ok := asyncFuncs[callee]; !ok || callee == sleep || callee == deadlockStub || callee == chanSendStub || callee == chanRecvStub {
 						continue
 					}
 					asyncCalls = append(asyncCalls, inst)
@@ -437,6 +441,26 @@ func (c *Compiler) markAsyncFunctions() (needsScheduler bool, err error) {
 		sw.AddCase(llvm.ConstInt(c.ctx.Int8Type(), 0, false), wakeup)
 		sw.AddCase(llvm.ConstInt(c.ctx.Int8Type(), 1, false), frame.cleanupBlock)
 		sleepCall.EraseFromParentAsInstruction()
+	}
+
+	// Transform calls to runtime.deadlockStub into coroutine suspends (without
+	// resume).
+	for _, deadlockCall := range getUses(deadlockStub) {
+		// deadlockCall must be a call instruction.
+		frame := asyncFuncs[deadlockCall.InstructionParent().Parent()]
+
+		// Exit coroutine.
+		c.builder.SetInsertPointBefore(deadlockCall)
+		continuePoint := c.builder.CreateCall(coroSuspendFunc, []llvm.Value{
+			llvm.ConstNull(c.ctx.TokenType()),
+			llvm.ConstInt(c.ctx.Int1Type(), 1, false), // final suspend
+		}, "")
+		c.splitBasicBlock(deadlockCall, llvm.NextBasicBlock(c.builder.GetInsertBlock()), "task.wakeup.dead")
+		c.builder.SetInsertPointBefore(deadlockCall)
+		sw := c.builder.CreateSwitch(continuePoint, frame.suspendBlock, 2)
+		sw.AddCase(llvm.ConstInt(c.ctx.Int8Type(), 0, false), frame.unreachableBlock)
+		sw.AddCase(llvm.ConstInt(c.ctx.Int8Type(), 1, false), frame.cleanupBlock)
+		deadlockCall.EraseFromParentAsInstruction()
 	}
 
 	// Transform calls to runtime.chanSendStub into channel send operations.

--- a/interp/frame.go
+++ b/interp/frame.go
@@ -310,6 +310,8 @@ func (fr *frame) evalBasicBlock(bb, incoming llvm.BasicBlock, indent string) (re
 			case callee.Name() == "runtime.makeInterface":
 				uintptrType := callee.Type().Context().IntType(fr.TargetData.PointerSize() * 8)
 				fr.locals[inst] = &LocalValue{fr.Eval, llvm.ConstPtrToInt(inst.Operand(0), uintptrType)}
+			case callee.Name() == "runtime.nanotime":
+				fr.locals[inst] = &LocalValue{fr.Eval, llvm.ConstInt(fr.Mod.Context().Int64Type(), 0, false)}
 			case strings.HasPrefix(callee.Name(), "runtime.print") || callee.Name() == "runtime._panic":
 				// This are all print instructions, which necessarily have side
 				// effects but no results.

--- a/interp/scan.go
+++ b/interp/scan.go
@@ -28,6 +28,9 @@ func (e *Eval) hasSideEffects(fn llvm.Value) *sideEffectResult {
 	case "runtime.alloc":
 		// Cannot be scanned but can be interpreted.
 		return &sideEffectResult{severity: sideEffectNone}
+	case "runtime.nanotime":
+		// Fixed value at compile time.
+		return &sideEffectResult{severity: sideEffectNone}
 	case "runtime._panic":
 		return &sideEffectResult{severity: sideEffectLimited}
 	}

--- a/src/reflect/value.go
+++ b/src/reflect/value.go
@@ -365,6 +365,25 @@ func (v Value) MapIndex(key Value) Value {
 	panic("unimplemented: (reflect.Value).MapIndex()")
 }
 
+func (v Value) MapRange() *MapIter {
+	panic("unimplemented: (reflect.Value).MapRange()")
+}
+
+type MapIter struct {
+}
+
+func (it *MapIter) Key() Value {
+	panic("unimplemented: (*reflect.MapIter).Key()")
+}
+
+func (it *MapIter) Value() Value {
+	panic("unimplemented: (*reflect.MapIter).Value()")
+}
+
+func (it *MapIter) Next() bool {
+	panic("unimplemented: (*reflect.MapIter).Next()")
+}
+
 func (v Value) Set(x Value) {
 	if !v.indirect {
 		panic("reflect: value is not addressable")

--- a/src/runtime/chan.go
+++ b/src/runtime/chan.go
@@ -41,6 +41,7 @@ const (
 
 func chanSendStub(caller *coroutine, ch *channel, _ unsafe.Pointer, size uintptr)
 func chanRecvStub(caller *coroutine, ch *channel, _ unsafe.Pointer, _ *bool, size uintptr)
+func deadlockStub()
 
 // chanSend sends a single value over the channel. If this operation can
 // complete immediately (there is a goroutine waiting for a value), it sends the

--- a/src/runtime/runtime.go
+++ b/src/runtime/runtime.go
@@ -84,9 +84,13 @@ func sleep(d int64) {
 	sleepTicks(timeUnit(d / tickMicros))
 }
 
+func nanotime() int64 {
+	return int64(ticks()) * tickMicros
+}
+
 //go:linkname now time.now
 func now() (sec int64, nsec int32, mono int64) {
-	mono = int64(ticks()) * tickMicros
+	mono = nanotime()
 	sec = mono / (1000 * 1000 * 1000)
 	nsec = int32(mono - sec*(1000*1000*1000))
 	return

--- a/src/runtime/runtime_wasm.go
+++ b/src/runtime/runtime_wasm.go
@@ -37,6 +37,11 @@ func putchar(c byte) {
 	resource_write(stdout, &c, 1)
 }
 
+//go:linkname setEventHandler syscall/js.setEventHandler
+func setEventHandler(fn func()) {
+	// TODO
+}
+
 //go:export go_scheduler
 func go_scheduler() {
 	scheduler()

--- a/testdata/channel.go
+++ b/testdata/channel.go
@@ -43,6 +43,10 @@ func main() {
 	}
 	println("sum(100):", sum)
 
+	// Test select
+	go selectDeadlock()
+	go selectNoOp()
+
 	// Allow goroutines to exit.
 	time.Sleep(time.Microsecond)
 }
@@ -92,4 +96,18 @@ func iterator(ch chan int, top int) {
 		ch <- i
 	}
 	close(ch)
+}
+
+func selectDeadlock() {
+	println("deadlocking")
+	select {}
+	println("unreachable")
+}
+
+func selectNoOp() {
+	println("select no-op")
+	select {
+	default:
+	}
+	println("after no-op")
 }

--- a/testdata/channel.txt
+++ b/testdata/channel.txt
@@ -19,3 +19,6 @@ sum: 25
 sum: 29
 sum: 33
 sum(100): 4950
+deadlocking
+select no-op
+after no-op

